### ISSUE TITLE
🐛 BUG: `Astro.fetchContent` can't be called in a page and a child component

### ISSUE THREAD

**Issue by tony-sull** [withastro/astro#2419]
### What version of `astro` are you using?

0.22.14

### What package manager are you using?

npm

### What operating system are you using?

Linux

### Describe the Bug

Astro's build will hang if `Astro.fetchContent` is used on the same directory by a page and one of it's child components

In the minimal repro, `astro build` hangs indefinitely as-is.  You can fix this by 1) commenting out one of the `Astro.fetchContent` calls in either `components/RelatedPosts.astro` or `pages/index.astro`, or 2) changing one of the `fetchContent` calls so they are searching different directories

This was originally hit by @rebelchris in a project he's migrating over to Astro

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-vuvgtd

**Comment by tony-sull** [comment]
Looks like `fetchContent` may be getting into a deadlock here.  Maybe the first call in the page has the directory, or the files, locked up for read access?

Not sure if this is something we can fix on our end or if it's actually a limitation outside Astro's control.  If nothing else maybe we can recognize this and give a build error with details on the issue and how to fix it (like only loading content in the page and passing that down as props to children)

**Comment by rebelchris** [comment]
Agreed with Tony

If it turns out to be non-fixable, a error would be appreciated.
Another solution for the end user is to use their own implementation of the glob.fetch case. (This seems to work for me atleast)

Passing down would also work, but depending on the use-case could be difficult.

**Comment by tusamni** [comment]
I believe I'm hitting this as well, though I can't be sure. I've got a related posts feature which calls fetchContent and then filters down to the related post slugs in the post markdown file.

I just figured it was slowing my build times down since I had to fetch all my posts every time a post was built. However, my builds are timing out on Netlify. I've only got 10 posts, I doubt that would cause a timeout on Netlify.
